<a href="https://colab.research.google.com/github/fedunya/Jupyter/blob/main/Jupyter_seminar3.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [1]:
import pandas as pd

## Задача 1  
### 1.1 Считать данные с помощью pandas. Вывести на экран первые 5 строк

In [2]:
df = pd.read_csv('./kc_house_data.csv', sep=',')
df.head()

id             date     price  bedrooms  bathrooms  sqft_living  \
0  7129300520  20141013T000000  221900.0         3       1.00         1180   
1  6414100192  20141209T000000  538000.0         3       2.25         2570   
2  5631500400  20150225T000000  180000.0         2       1.00          770   
3  2487200875  20141209T000000  604000.0         4       3.00         1960   
4  1954400510  20150218T000000  510000.0         3       2.00         1680   

   sqft_lot  floors  waterfront  view  ...  grade  sqft_above  sqft_basement  \
0      5650     1.0           0     0  ...      7        1180              0   
1      7242     2.0           0     0  ...      7        2170            400   
2     10000     1.0           0     0  ...      6         770              0   
3      5000     1.0           0     0  ...      7        1050            910   
4      8080     1.0           0     0  ...      8        1680              0   

   yr_built  yr_renovated  zipcode      lat     long  sqft_living15  \
0      1955             0    98178  47.5112 -122.257           1340   
1      1951          1991    98125  47.7210 -122.319           1690   
2      1933             0    98028  47.7379 -122.233           2720   
3      1965             0    98136  47.5208 -122.393           1360   
4      1987             0    98074  47.6168 -122.045           1800   

   sqft_lot15  
0        5650  
1        7639  
2        8062  
3        5000  
4        7503  

[5 rows x 21 columns]

### 1.2 Создать новый признак delta_renovated, который будет содержать разницу в годах между годом реновацией дома и годом постройки дома. Если реновации дома не было, то в новом признаке поставьте 0

In [3]:
df['delta_renovated'] = df['yr_renovated'] - df['yr_built']
df.loc[df['delta_renovated'] < 0, 'delta_renovated'] = 0
df.head()

id             date     price  bedrooms  bathrooms  sqft_living  \
0  7129300520  20141013T000000  221900.0         3       1.00         1180   
1  6414100192  20141209T000000  538000.0         3       2.25         2570   
2  5631500400  20150225T000000  180000.0         2       1.00          770   
3  2487200875  20141209T000000  604000.0         4       3.00         1960   
4  1954400510  20150218T000000  510000.0         3       2.00         1680   

   sqft_lot  floors  waterfront  view  ...  sqft_above  sqft_basement  \
0      5650     1.0           0     0  ...        1180              0   
1      7242     2.0           0     0  ...        2170            400   
2     10000     1.0           0     0  ...         770              0   
3      5000     1.0           0     0  ...        1050            910   
4      8080     1.0           0     0  ...        1680              0   

   yr_built  yr_renovated  zipcode      lat     long  sqft_living15  \
0      1955             0    98178  47.5112 -122.257           1340   
1      1951          1991    98125  47.7210 -122.319           1690   
2      1933             0    98028  47.7379 -122.233           2720   
3      1965             0    98136  47.5208 -122.393           1360   
4      1987             0    98074  47.6168 -122.045           1800   

   sqft_lot15  delta_renovated  
0        5650                0  
1        7639               40  
2        8062                0  
3        5000                0  
4        7503                0  

[5 rows x 22 columns]

### 1.3 Создайте признаки года продажи, месяца продажи

In [4]:
df['yr_sale'] = df['date'].apply(lambda x: x[:4])
df['month_sale'] = df['date'].apply(lambda x: x[4:6])
df.head()

id             date     price  bedrooms  bathrooms  sqft_living  \
0  7129300520  20141013T000000  221900.0         3       1.00         1180   
1  6414100192  20141209T000000  538000.0         3       2.25         2570   
2  5631500400  20150225T000000  180000.0         2       1.00          770   
3  2487200875  20141209T000000  604000.0         4       3.00         1960   
4  1954400510  20150218T000000  510000.0         3       2.00         1680   

   sqft_lot  floors  waterfront  view  ...  yr_built  yr_renovated  zipcode  \
0      5650     1.0           0     0  ...      1955             0    98178   
1      7242     2.0           0     0  ...      1951          1991    98125   
2     10000     1.0           0     0  ...      1933             0    98028   
3      5000     1.0           0     0  ...      1965             0    98136   
4      8080     1.0           0     0  ...      1987             0    98074   

       lat     long  sqft_living15  sqft_lot15  delta_renovated  yr_sale  \
0  47.5112 -122.257           1340        5650                0     2014   
1  47.7210 -122.319           1690        7639               40     2014   
2  47.7379 -122.233           2720        8062                0     2015   
3  47.5208 -122.393           1360        5000                0     2014   
4  47.6168 -122.045           1800        7503                0     2015   

   month_sale  
0          10  
1          12  
2          02  
3          12  
4          02  

[5 rows x 24 columns]

### 1.4 Удалите признаки date, zipcode, lat, long

In [5]:
df.drop(columns=['date', 'zipcode', 'lat', 'long'], inplace=True)
# df.drop(['date', 'zipcode', 'lat', 'long'], axis=1, inplace=True) - второй вариант удаления
df.head()

id     price  bedrooms  bathrooms  sqft_living  sqft_lot  floors  \
0  7129300520  221900.0         3       1.00         1180      5650     1.0   
1  6414100192  538000.0         3       2.25         2570      7242     2.0   
2  5631500400  180000.0         2       1.00          770     10000     1.0   
3  2487200875  604000.0         4       3.00         1960      5000     1.0   
4  1954400510  510000.0         3       2.00         1680      8080     1.0   

   waterfront  view  condition  grade  sqft_above  sqft_basement  yr_built  \
0           0     0          3      7        1180              0      1955   
1           0     0          3      7        2170            400      1951   
2           0     0          3      6         770              0      1933   
3           0     0          5      7        1050            910      1965   
4           0     0          3      8        1680              0      1987   

   yr_renovated  sqft_living15  sqft_lot15  delta_renovated yr_sale month_sale  
0             0           1340        5650                0    2014         10  
1          1991           1690        7639               40    2014         12  
2             0           2720        8062                0    2015         02  
3             0           1360        5000                0    2014         12  
4             0           1800        7503                0    2015         02

## Задача 2  
Создайте датафрейм с клиентами:

In [6]:
clients = pd.DataFrame({
'client_id': [1459, 4684, 3498, 3942, 4535, 2308, 2866, 2765, 1472, 4236, 2295,
              939, 3840, 280, 20, 4332, 3475, 4213, 3113, 4809, 2134, 2242,
              2068, 4929, 1384, 1589, 3317, 2260, 1727, 1764, 1611, 1474],
'house_id': [8965450190, 6823100225, 5104540330, 2131701075, 1522700060,
             1189000207, 6821600300, 7137950720, 9510920050, 6131600255,
             5428000070, 1788800910, 8100400160, 3123049142, 6306800010,
             5083000375, 7920100025, 1951600150, 809001400, 339600110,
             1622049154, 1099600250, 8563000110, 2768100205, 3995700435,
             8861700030, 3303980210, 7731100066, 8146100580, 825069097,
             3889100029, 9524100196]
})

### 2.1 Присоедините к таблице clients данные по домам через метод join

In [7]:
df_id = df.set_index('id')
clients_id = clients.set_index('house_id')
clients_id.join(df_id).head()

client_id     price  bedrooms  bathrooms  sqft_living  sqft_lot  \
339600110        4809  395000.0         3       2.50         1610      3755   
809001400        3113  925000.0         3       1.00         1630      3200   
825069097        1764  770000.0         3       2.50         2650     40705   
1099600250       2242  260000.0         3       1.75         1710      6400   
1189000207       2308  387000.0         2       2.50         1170      1394   

            floors  waterfront  view  condition  grade  sqft_above  \
339600110      2.0           0     0          3      7        1610   
809001400      1.5           0     0          4      8        1630   
825069097      2.0           0     0          3      9        2650   
1099600250     1.0           0     0          4      7        1240   
1189000207     2.0           0     0          3      8        1170   

            sqft_basement  yr_built  yr_renovated  sqft_living15  sqft_lot15  \
339600110               0      1987             0           1300        3823   
809001400               0      1912             0           1710        3600   
825069097               0      1994             0           2550       42625   
1099600250            470      1976             0           1600        6400   
1189000207              0      2001             0           1250        3136   

            delta_renovated yr_sale month_sale  
339600110                 0    2014         09  
809001400                 0    2014         09  
825069097                 0    2014         06  
1099600250                0    2014         12  
1189000207                0    2014         10

### 2.2 Присоедините к таблице clients данные по домам через метод merge. Это нужно, чтобы понимать, какие дома покупались клиентами house_id - это индексы датафрейма с домами

In [8]:
merged = df.merge(clients, left_on='id', right_on='house_id')
merged.head()

id     price  bedrooms  bathrooms  sqft_living  sqft_lot  floors  \
0  8100400160  700000.0         3       2.25         2330     11424     2.0   
1  7920100025  450000.0         2       1.00          740      5100     1.0   
2   825069097  770000.0         3       2.50         2650     40705     2.0   
3  5083000375  170000.0         3       1.00         1310      9529     1.0   
4  5083000375  235000.0         3       1.00         1310      9529     1.0   

   waterfront  view  condition  ...  sqft_basement  yr_built  yr_renovated  \
0           0     0          4  ...              0      1984             0   
1           0     0          4  ...              0      1947             0   
2           0     0          3  ...              0      1994             0   
3           0     0          3  ...              0      1956             0   
4           0     0          3  ...              0      1956             0   

   sqft_living15  sqft_lot15  delta_renovated  yr_sale  month_sale client_id  \
0           2050       11448                0     2015          04      3840   
1            920        5100                0     2015          04      3475   
2           2550       42625                0     2014          06      1764   
3           1330        9529                0     2014          10      4332   
4           1330        9529                0     2015          03      4332   

     house_id  
0  8100400160  
1  7920100025  
2   825069097  
3  5083000375  
4  5083000375  

[5 rows x 22 columns]

## Задача 3  
Составьте несколько сводных таблиц

### 3.1 Найдите среднюю стоимость домов в зависимости от количества спален Отсортируйте от меньшей стоимости к большей

In [9]:
df.groupby('bedrooms').agg({'price': 'mean'}).sort_values('price')

price
bedrooms              
1         3.176429e+05
2         4.013727e+05
0         4.095038e+05
3         4.662321e+05
11        5.200000e+05
4         6.354195e+05
33        6.400000e+05
5         7.865998e+05
10        8.193333e+05
6         8.255206e+05
9         8.939998e+05
7         9.511847e+05
8         1.105077e+06

### 3.2 Найдите минимальную, среднюю и максимальную стоимости домов в зависимости от состояния дома

In [10]:
df.groupby('condition').agg({'price': ['min', 'mean', 'max']})

price                          
                min           mean        max
condition                                    
1           78000.0  334431.666667  1500000.0
2           80000.0  327287.145349  2555000.0
3           75000.0  542012.578148  7062500.0
4           89000.0  521200.390033  7700000.0
5          110000.0  612418.089359  3650000.0

### 3.3 Постройте таблицу с подсчетом количества домов в данных в зависимости от вида на набережную и оценкой вида

In [12]:
df.pivot_table(index='waterfront', columns='view', values='id', aggfunc='count', fill_value=0)

view            0    1    2    3    4
waterfront                           
0           19489  331  955  491  184
1               0    1    8   19  135

### 3.4 Каких домов в зависимости от этажности и количества спален больше?

In [15]:
df.pivot_table(index='floors', columns='bedrooms', values='id', aggfunc='count', fill_value=0)

bedrooms  0    1     2     3     4    5    6   7   8   9   10  11  33
floors                                                               
1.0        4  162  1951  5455  2383  605  104   9   5   0   1   0   1
1.5        0   21   182   786   698  185   30   7   1   0   0   0   0
2.0        6   12   497  3118  3682  775  119  19   6   4   2   1   0
2.5        0    1     5    56    58   23   14   2   0   2   0   0   0
3.0        2    3   123   405    61   13    5   1   0   0   0   0   0
3.5        1    0     2     4     0    0    0   0   1   0   0   0   0

Больше всего 1-этажных домов с 3-мя спальнями - 5455

### 3.5 Постройте таблицу с подсчетом медианной стоимости домов в данных в зависимости от состояния дома и оценки дома

In [19]:
pd.crosstab(index=df['condition'], columns=df['grade'], values=df['price'], aggfunc='median').fillna(0)

grade            1         3         4         5         6         7   \
condition                                                               
1          142000.0       0.0  150000.0  190000.0  255000.0  403500.0   
2               0.0  280000.0  145000.0  180000.0  235000.0  305000.0   
3               0.0   75000.0  205000.0  234475.0  265000.0  357500.0   
4               0.0       0.0  238525.0  229975.0  295000.0  390000.0   
5               0.0  262000.0       0.0  227450.0  285475.0  456000.0   

grade            8          9          10         11         12         13  
condition                                                                   
1          932500.0        0.0        0.0        0.0        0.0        0.0  
2          429000.0   715000.0  1752500.0        0.0        0.0        0.0  
3          485000.0   689000.0   890000.0  1209500.0  1807500.0  2888000.0  
4          571250.0   823500.0  1030000.0  1685000.0  2125000.0  5750000.0  
5          696000.0  1078000.0  1650000.0  2050000.0  1990000.0        0.0